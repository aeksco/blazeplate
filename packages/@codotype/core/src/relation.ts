import { UUID } from "./uuid";
import { TokenPluralization } from "./token";
import { SchemaCreators } from "./schema";

// // // //

/**
 * RelationType
 * Defines an enum of available RelationsTypes
 */
export type RelationType =
    | "BELONGS_TO"
    | "HAS_ONE"
    | "HAS_MANY"
    | "TO_ONE"
    | "TO_MANY"
    | "HAS_AND_BELONGS_TO_MANY";
export enum RelationTypes {
    BELONGS_TO = "BELONGS_TO",
    HAS_ONE = "HAS_ONE",
    HAS_MANY = "HAS_MANY",
    TO_ONE = "TO_ONE",
    TO_MANY = "TO_MANY",
    HAS_AND_BELONGS_TO_MANY = "HAS_AND_BELONGS_TO_MANY",
}

/**
 * RelationInput
 * Defines the RelationInput interface
 * TODO - add support for addons here -> see AttributeAddon for context
 */
export interface RelationInput {
    id: UUID;
    type: RelationType;
    required: boolean; // TODO - keep this for now - should this be handled as an Addon? YES.
    source: SchemaCreators;
    sourceSchemaID: UUID;
    destinationSchemaID: UUID;
    sourceSchemaAlias: string;
    destinationSchemaAlias: string;
}

// Relation constants
export const DEFAULT_RELATION_INPUT: RelationInput = {
    id: "",
    type: null,
    required: false,
    source: SchemaCreators.user,
    sourceSchemaID: "",
    destinationSchemaID: "",
    sourceSchemaAlias: "",
    destinationSchemaAlias: "",
};

// // // //

/**
 * Relation
 * Defines the Reference interface
 * TODO - add support for addons here -> see AttributeAddon for context
 */
export interface Relation {
    id: UUID;
    type: RelationType;
    sourceRelationId: UUID;
    sourceSchemaId: UUID;
    destinationSchemaId: UUID;
    identifiers: {
        source: {
            canonical: TokenPluralization;
            alias: TokenPluralization;
        };
        destination: {
            canonical: TokenPluralization;
            alias: TokenPluralization;
        };
    };
}
