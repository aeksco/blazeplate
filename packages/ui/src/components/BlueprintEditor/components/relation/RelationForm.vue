<template>
  <b-modal
    lazy
    :visible="showingModal"
    :title="'New Relation'"
    ok-title='Submit'
    ok-variant='success'
    cancel-title='Cancel'
    cancel-variant='light'
    @ok="submit(newModel)"
    @hide="showModal(false)"
  >

    <b-form>
      <p class="mb-2 form-text text-muted">TODO - add relation form here</p>
      <input type="" v-model="newModel.label" name="">
    </b-form>

  </b-modal>
</template>

<script>
import { mapGetters, mapActions, mapMutations } from 'vuex'

export default {
  name: 'RelationForm',
  computed: mapGetters({
    showingModal: 'editor/schema/relation/modals/form/showing',
    newModel: 'editor/schema/relation/collection/newModel'
  }),
  methods: {
    ...mapActions({
      createPersonModel: 'editor/schema/relation/collection/create',
      updateParentSchemaRelations: 'editor/schema/updateRelations'
    }),
    ...mapMutations({
      showModal: 'editor/schema/relation/modals/form/showing',
      setNewPersonModel: 'editor/schema/relation/collection/newModel'
    }),
    submit (newModel) {
      this.setNewPersonModel(newModel)
      this.createPersonModel(),
      this.updateParentSchemaRelations()
    }
  }
}
</script>
