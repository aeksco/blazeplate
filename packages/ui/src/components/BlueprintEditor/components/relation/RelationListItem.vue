<template>
  <li class='list-group-item'>
    {{ model }}
  </li>
</template>

<script>
export default {
  name: 'AttributeListItem',
  props: ['model']
}
</script>
