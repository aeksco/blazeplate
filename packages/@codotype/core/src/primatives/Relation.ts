import { v4 as uuidv4 } from "uuid";
import { RelationTypes, RelationInput } from "../relation";
import { SchemaCreators } from "../schema";

interface RelationBuilderParams {
    id?: string;
    type: RelationTypes;
    sourceSchemaID: string;
    destinationSchemaID: string;
    required?: boolean;
    source?: SchemaCreators;
    sourceSchemaAlias?: string;
    destinationSchemaAlias?: string;
    internalNote?: string;
}

export class RelationBuilder implements RelationInput {
    id: string = uuidv4();
    type: RelationTypes;
    sourceSchemaID: string;
    destinationSchemaID: string;
    internalNote: string = "";
    required: boolean = false;
    source: SchemaCreators = SchemaCreators.user;
    sourceSchemaAlias: string = "";
    destinationSchemaAlias: string = "";

    constructor(params: RelationBuilderParams) {
        this.type = params.type;
        this.sourceSchemaID = params.sourceSchemaID;
        this.destinationSchemaID = params.destinationSchemaID;
        this.internalNote = params.internalNote;

        this.id = params.id || this.id;
        this.required = params.required || this.required;
        this.source = params.source || this.source;
        this.sourceSchemaAlias =
            params.sourceSchemaAlias || this.sourceSchemaAlias;
        this.destinationSchemaAlias =
            params.destinationSchemaAlias || this.destinationSchemaAlias;
    }
}
