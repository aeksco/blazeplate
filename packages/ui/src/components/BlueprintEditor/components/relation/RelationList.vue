<template>
  <b-row>
    <b-col lg=12>

      <div class='card' id='relation-detail'>

        <div class="card-header d-flex justify-content-between align-items-center">

          <span>
            <i class="fa fa-link mr-1"></i>
            Relationships
          </span>

          <NewModalButton
            id="add-relation-button"
            vuexModule="editor/schema/relation/modals/form/showing"
          />

        </div>

        <b-list-group flush v-if="collection[0]">
          <RelationListItem v-for="item in collection" :key="item.id" :model="item" />
        </b-list-group>

        <!-- NOTES FROM OLD VERSION -->
        <!-- <ul v-if="!relations.length && !referencedBy.length"> -->
        <b-list-group flush v-else>
          <b-list-group-item class="text-center text-primary">
            <i class="fa fa-lg fa-info-circle"></i>
            <p class="mb-0 mt-1">
              <!-- TODO - Click to add your first Relation -->
              Relations define associations between different models
            </p>
          </b-list-group-item>
        </b-list-group>

      </div>

    </b-col>
  </b-row>
</template>

<script>
import { mapGetters } from 'vuex'
import NewModalButton from '../NewModalButton'
import RelationListItem from './RelationListItem'

export default {
  name: 'RelationList',
  components: {
    NewModalButton,
    RelationListItem
  },
  computed: mapGetters({
    collection: 'editor/schema/relation/collection/items'
  })
}
</script>
