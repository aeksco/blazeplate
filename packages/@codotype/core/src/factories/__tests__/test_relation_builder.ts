import { RelationBuilder } from "../Relation";
import { RelationTypes } from "../../relation";

// Mocks UUID module
jest.mock("uuid", () => ({
    __esModule: true,
    v4: () => "1234-5678-1234-5678",
}));

// // // //

describe("factory", () => {
    test("works", () => {
        const relationBuilder = new RelationBuilder({
            type: RelationTypes.BELONGS_TO,
            destinationSchemaId: "SCHEMA_01",
        });

        expect(relationBuilder).toMatchSnapshot();
    });
});
