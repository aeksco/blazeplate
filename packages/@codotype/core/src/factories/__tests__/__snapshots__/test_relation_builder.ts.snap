// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`factory works 1`] = `
RelationBuilder {
  "destinationSchemaAlias": "",
  "destinationSchemaID": undefined,
  "id": "1234-5678-1234-5678",
  "required": false,
  "source": "user",
  "sourceSchemaAlias": "",
  "sourceSchemaID": undefined,
  "type": "BELONGS_TO",
}
`;
